src/ophys_etl/modules/segmentation/graph_utils/edge_attributes.py

```diff
@@ -0,0 +1,381 @@
+import h5py
+import networkx as nx
+import numpy as np
+from pathlib import Path
+from typing import Optional, Union, Tuple, List
+from scipy.spatial.distance import cdist
+from scipy.ndimage import gaussian_filter
+
+
+def normalize_graph(graph: nx.Graph,
+                    attribute_name: str,
+                    sigma: float = 30.0,
+                    new_attribute_name: Optional[str] = None) -> nx.Graph:
+    """normalizes edge weights by a local Gaussian filter
+    of size sigma.
+
+    Parameters
+    ----------
+    graph: nx.Graph
+        a graph with edge weights
+    sigma: float
+        passed to scipy.ndimage.gaussian_filter as 'sigma'
+    attribute_name: str
+        the name of the edge attribute to normalize
+    new_attribute_name: str
+        the name of the new edge attribute. If 'None' will be
+        <attribute_name>_normalized
+
+    Returns
+    -------
+    graph: nx.Graph
+        graph with the new edge attribute
+
+    """
+    if new_attribute_name is None:
+        new_attribute_name = attribute_name + "_normalized"
+
+    # create an image that is the average edge attribute per node
+    coords = np.array(list(graph.nodes))
+    shape = tuple(coords.max(axis=0) + 1)
+    avg_attr = np.zeros(shape, dtype='float')
+    for node in graph.nodes:
+        n = len(list(graph.neighbors(node)))
+        avg_attr[node[0], node[1]] = \
+            graph.degree(node, weight=attribute_name) / n
+
+    # Gaussian filter the average image
+    avg_attr = gaussian_filter(avg_attr, sigma, mode='nearest')
+    edge_values = dict()
+    for i, v in graph.edges.items():
+        local = 0.5 * (avg_attr[i[0][0], i[0][1]] + avg_attr[i[1][0], i[1][0]])
+        edge_values[i] = {new_attribute_name: v[attribute_name] / local}
+
+    nx.set_edge_attributes(graph, edge_values)
+    return graph
+
+
+def add_pearson_edge_attributes(graph: nx.Graph,
+                                video_path: Path,
+                                attribute_name: str = "Pearson") -> nx.Graph:
+    """adds an attribute to each edge which is the Pearson correlation
+    coefficient between the traces of the two pixels (nodes) associated
+    with that edge.
+
+    Parameters
+    ----------
+    graph: nx.Graph
+        a graph with nodes like (row, col) and edges connecting them
+    video_path: Path
+        path to an hdf5 video file, assumed to have a dataset "data"
+        nframes x nrow x ncol
+    attribute_name: str
+        name set on each edge for this calculated value
+
+    Returns
+    -------
+    new_graph: nx.Graph
+        an undirected networkx graph, with attribute added to edges
+
+    """
+    new_graph = nx.Graph()
+    # copies over node attributes
+    new_graph.add_nodes_from(graph.nodes(data=True))
+
+    # load the section of data that encompasses this graph
+    rows, cols = np.array(graph.nodes).T
+    with h5py.File(video_path, "r") as f:
```

Comment:
Function should probably not accept a `video_path` and instead should accept an array of video data to separate out IO concerns.

Also, has this been tried with extremely long/big movies? Loading the whole thing in all at once instead of over chunks of frames may cause problems?

Reply:
re: video_path, I'll look into that change.
re: long movies, it works, but, it is slow. chunking the video in time is not ideal, since the metrics operate on time series, and recombining metrics from time chunks is not ideal - i.e. one could calculate Pearson in [time_a, time_b] and [time_b, time_c] but, can't reconstruct [time_a, time_c] from those (at least easily and in a "works generally with any metric sort of way") Instead, I've implemented partitioning the graph in space, see `partition.py` which does generalize well and preserves individual time series.

Reply:
re: video_path
this is a parallelization strategy choice. jobs get the video path, open, and extract only the data they need.
the jobs figure out what data they need based on other arguments (the graph partition they are passed)
this is preferable to passing around big chunks of data, or open `h5py.File` objects, which probably run into problems getting serialized properly for multiprocessing jobs.